### ISSUE TITLE
RuntimeError: can't start new thread

### ISSUE THREAD

**Issue by RichardCorbett** [nanoporetech/medaka#297]
Hi there,

I am running this command having given it 8 CPUs and 64Gb of RAM on my slurm cluster.
I just installed this through conda and it is running version 1.3.2.

I am testing this on very very shallow data just to get it up and running.   Please let me know if this is the possible cause of the issue.

```
Caused by:
  Process `medaka_1_3_2_wf:medaka_1_3_2_variant (D74501_4)` terminated with an error exit status (1)

Command executed:
  export OMP_NUM_THREADS=1
  
  medaka_variant -f hg19a_phage_lambda.fa         -i D74501.sorted.bam         -P 0         -l         -r 4 
  
Command exit status:
  1

Command output:
  
  ++ Preparing data
  - Checking program versions
  Program    Version    Required   Pass     
  bcftools   1.12       1.11       True     
  bgzip      1.12       1.11       True     
  minimap2   2.18       2.11       True     
  samtools   1.12       1.11       True     
  tabix      1.12       1.11       True     
  + Extracting region from bam
  - Extracting regions 4 from bam /projects/rcorbettprj2/2019FFPE/scratch/promethion/tmp/work/c9/41e4a29e13b2221e7bd360cef6e68e/D74501.sorted.bam 
  + Indexing bam
  - Running samtools index on D74501.sorted.bam.
    Bam index in medaka_variant/D74501.sorted.bam.bai.
  
  ++ Running first pass to haplotag reads.
  + Running consensus calling
  - Running medaka consensus with D74501.sorted.bam --model r941_prom_snp_g360 
  * Failed to run medaka consensus.

Command error:
  [12:16:29 - PWorker] Processed 1 batches
  [12:16:29 - PWorker] All done, 0 remainder regions.
  [12:16:29 - Sampler] Initializing sampler for consensus of region 4:7611018-7615023.
  [12:16:29 - PWorker] Running inference for 0.0M draft bases.
  [12:16:29 - Feature] Processed 4:7611018.0-7615022.0 (median depth 1.0)
  [12:16:29 - Sampler] Took 0.27s to make features.
  [12:16:32 - PWorker] Processed 1 batches
  [12:16:32 - PWorker] All done, 0 remainder regions.
  [12:16:32 - Sampler] Initializing sampler for consensus of region 4:7823305-7827740.
  [12:16:32 - PWorker] Running inference for 0.0M draft bases.
  [12:16:32 - Feature] Processed 4:7823305.0-7827739.0 (median depth 1.0)
  [12:16:32 - Sampler] Took 0.29s to make features.
  [12:16:35 - PWorker] Processed 1 batches
  [12:16:35 - PWorker] All done, 0 remainder regions.
  [12:16:35 - Sampler] Initializing sampler for consensus of region 4:8991000-8997520.
  [12:16:35 - PWorker] Running inference for 0.0M draft bases.
  [12:16:35 - Feature] Processed 4:8991000.0-8997519.0 (median depth 1.0)
  [12:16:35 - Sampler] Took 0.32s to make features.
  [12:16:38 - PWorker] Processed 1 batches
  [12:16:38 - PWorker] All done, 0 remainder regions.
  [12:16:38 - Sampler] Initializing sampler for consensus of region 4:9016115-9017777.
  [12:16:38 - PWorker] Running inference for 0.0M draft bases.
  [12:16:38 - Feature] Processed 4:9016115.0-9017776.0 (median depth 1.0)
  [12:16:38 - Sampler] Took 0.21s to make features.
  [12:16:40 - PWorker] Processed 1 batches
  [12:16:40 - PWorker] All done, 0 remainder regions.
  [12:16:40 - Sampler] Initializing sampler for consensus of region 4:9141880-9150713.
  [12:16:40 - PWorker] Running inference for 0.0M draft bases.
  [12:16:41 - Feature] Processed 4:9141880.0-9150712.0 (median depth 1.0)
  [12:16:41 - Sampler] Took 0.37s to make features.
  [12:16:45 - PWorker] Processed 1 batches
  [12:16:45 - PWorker] All done, 0 remainder regions.
  [12:16:45 - Sampler] Initializing sampler for consensus of region 4:9152935-9157013.
  [12:16:45 - MdlStore] ModelStore exception <class 'RuntimeError'>
  Traceback (most recent call last):
    File "/home/rcorbett/miniconda3/envs/medaka/bin/medaka", line 11, in <module>
      sys.exit(main())
    File "/home/rcorbett/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/medaka.py", line 688, in main
      args.func(args)
    File "/home/rcorbett/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/prediction.py", line 181, in predict
      new_remainders = run_prediction(
    File "/home/rcorbett/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/prediction.py", line 25, in run_prediction
      loader = DataLoader(
    File "/home/rcorbett/miniconda3/envs/medaka/lib/python3.8/site-packages/medaka/prediction.py", line 254, in __init__
      self._bthread.start()
    File "/home/rcorbett/miniconda3/envs/medaka/lib/python3.8/threading.py", line 852, in start
  [12:16:45 - Feature] Processed 4:9152935.0-9157012.0 (median depth 1.0)
  [12:16:45 - Sampler] Took 0.00s to make features.
      _start_new_thread(self._bootstrap, ())
  RuntimeError: can't start new thread
```

Do you have any suggestions about what could be happening?

**Comment by RichardCorbett** [comment]
I ran the same code on a 30X bam and didn't get any errors, so it is looking as though the error was specific to the shallow test I was performing.

**Comment by cjw85** [comment]
We've not encountered this. It looks to me like some worker threads are not being cleaned up in a timely manner and you are reaching an OX thread limit. From the above I can see that medaka is processing lots of small low depth contigs, this would certainly cause it to want to start (and hopefully stop) a lot of threads in quick succession.

I've quickly reviewed the code and could not see anything suspicious. I will try to contrive some test data to see if I can reproduce this behaviour.

**Comment by cjw85** [comment]
I believe I've found the issue here. It appears to be a bug, or at least undesirable behaviour, within the tensorflow library that medaka uses. I have found a way to work around the issue; we will make a new release shortly.

**Comment by cjw85** [comment]
A fix for this will be available in version 1.3.4.
